src/v/iceberg/avro_decimal.h

```diff
@@ -0,0 +1,77 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+#pragma once
+
+#include "bytes/iobuf.h"
+
+#include <absl/numeric/int128.h>
+
+#include <array>
+namespace iceberg {
+/**
+ * Converts a decimal into an array of bytes (big endian)
+ */
+
+inline std::array<uint8_t, 16> encode_avro_decimal(absl::int128 decimal) {
+    auto high_half = absl::Uint128High64(decimal);
+    auto low_half = absl::Uint128Low64(decimal);
+
+    std::array<uint8_t, 16> decimal_bytes{};
+    decimal_bytes.fill(0);
+
+    for (int i = 0; i < 8; i++) {
+        // NOLINTNEXTLINE(hicpp-signed-bitwise)
+        decimal_bytes.at(i) = (low_half >> (i * 8)) & 0xFF;
+        // NOLINTNEXTLINE(hicpp-signed-bitwise)
+        decimal_bytes.at(i + 8) = (high_half >> (i * 8)) & 0xFF;
+    }
+
+    return decimal_bytes;
+}
+/**
+ * Converts an array of big endian encoded bytes to an absl::int128
+ */
+inline absl::int128 decode_avro_decimal(std::array<uint8_t, 16> bytes) {
+    int64_t high_half = 0;
+    uint64_t low_half = 0;
+    for (size_t i = 0; i < 8; i++) {
+        low_half |= static_cast<uint64_t>(bytes.at(i)) << i * 8;
+        // NOLINTNEXTLINE(hicpp-signed-bitwise)
+        high_half |= static_cast<int64_t>(bytes.at(8 + i)) << i * 8;
+    }
+
+    return absl::MakeInt128(high_half, low_half);
+}
+
+inline iobuf avro_decimal_to_iobuf(absl::int128 decimal, size_t max_size) {
+    if (max_size > 16) {
+        throw std::invalid_argument(
+          "Decimal iobuf can not be larger than 16 bytes");
+    }
+    auto bytes = encode_avro_decimal(decimal);
+    iobuf buf;
+    buf.append(
+      // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+      reinterpret_cast<const char*>(bytes.data()),
+      std::min(bytes.size(), max_size));
+    return buf;
+}
+
+inline absl::int128 iobuf_to_avro_decimal(iobuf buf) {
+    if (buf.size_bytes() > 16) {
+        throw std::invalid_argument(
+          "Decimal iobuf can not be larger than 16 bytes");
+    }
+    std::array<uint8_t, 16> bytes{};
+    bytes.fill(0);
+    iobuf::iterator_consumer it(buf.cbegin(), buf.cend());
+    it.consume_to(buf.size_bytes(), bytes.data());
+    return decode_avro_decimal(bytes);
+}
```

Comment:
Can this be simplified with `iobuf_to_bytes`

https://github.com/redpanda-data/redpanda/blob/70d040413dbc47f7f5f8f0bc651158b1ef60d4f6/src/v/bytes/bytes.h#L213